src/vs/workbench/services/extensions/worker/extensionHostWorker.ts

```diff
@@ -219,18 +219,24 @@ function connectToRenderer(protocol: IMessagePassingProtocol): Promise<IRenderer
 
 let onTerminate = (reason: string) => nativeClose();
 
-export function create(): void {
-	const res = new ExtensionWorker();
+export function create(): { onmessage: (message: any) => void } {
 	performance.mark(`code/extHost/willConnectToRenderer`);
-	connectToRenderer(res.protocol).then(data => {
-		performance.mark(`code/extHost/didWaitForInitData`);
-		const extHostMain = new ExtensionHostMain(
-			data.protocol,
-			data.initData,
-			hostUtil,
-			null,
-		);
-
-		onTerminate = (reason: string) => extHostMain.terminate(reason);
-	});
+	const res = new ExtensionWorker();
+
+	return {
+		onmessage(messagePorts: ReadonlyMap<string, MessagePort>) {
```

Comment:
I was initially wondering if you can use the established `MessagePort` to send more message ports? I would keep the iframes a little simpler but I guess it makes the ExtensionWorker more complex (since it would receive the ports or have its ports being passed in...). Tho, it took me way to long to see/understand the `onmessage` here

Reply:
yeah, I believe it's OK like that